File: data-structures/non-linear/graph/Graph.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <queue>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

struct Node {
  string value;
  vector<pair<string, int>> adj;
};

class Graph {
  void DFSUtil(string &v, unordered_set<string> &visited) {
    visited.insert(v);
    cout << v << " ";

    for (auto &p : nodes[v].adj) {
      if (visited.find(p.first) == visited.end()) {
        DFSUtil(p.first, visited);
      }
    }
  }

  string findParent(string v, unordered_map<string, string> &parent) {
    if (parent[v] != v) {
      parent[v] = findParent(parent[v], parent);
    }

    return parent[v];
  }

  void unionSet(string u, string v, unordered_map<string, string> &parent,
                unordered_map<string, int> &rank) {
    u = findParent(u, parent);
    v = findParent(v, parent);

    if (u != v) {
      if (rank[u] < rank[v]) {
        parent[u] = v;
      } else if (rank[u] > rank[v]) {
        parent[v] = u;
      } else {
        parent[v] = u;
        rank[u]++;
      }
    }
  }

public:
  unordered_map<string, Node> nodes;
  vector<tuple<int, string, string>> edges;
  unordered_set<string> vertices;

  void insertNode(string &e) {
    if (nodes.find(e) == nodes.end()) {
      Node node;

      node.value = e;
      nodes[e] = node;
      vertices.insert(e);
    }
  }

  void insertEdge(string &v1, string &v2, int w) {
    if (nodes.find(v1) == nodes.end() || nodes.find(v2) == nodes.end()) {
      cout << "Vertex not found\n";
      return;
    }

    for (auto &p : nodes[v1].adj) {
      if (p.first == v2) {
        cout << "Edge " << v1 << "-" << v2 << " already exists\n";
        return;
      }
    }

    nodes[v1].adj.push_back({v2, w});
    nodes[v2].adj.push_back({v1, w});
    edges.push_back({w, v1, v2});
  }

  void DFS(string &start) {
    unordered_set<string> visited;

    DFSUtil(start, visited);
    cout << "\n";
  }

  void BFS(string &start) {
    unordered_set<string> visited;
    queue<string> q;

    q.push(start);
    visited.insert(start);

    while (!q.empty()) {
      string node = q.front();

      q.pop();
      cout << node << " ";

      for (auto &p : nodes[node].adj) {
        if (visited.find(p.first) == visited.end()) {
          visited.insert(p.first);
          q.push(p.first);
        }
      }
    }

    cout << "\n";
  }

  void Prim(string &start) {
    if (nodes.find(start) == nodes.end()) {
      cout << "Invalid start vertex\n";
      return;
    }

    unordered_set<string> visited;

    visited.insert(start);

    using Edge = tuple<int, string, string>;
    priority_queue<Edge, vector<Edge>, greater<Edge>> pq;

    for (auto &p : nodes[start].adj) {
      pq.push({p.second, start, p.first});
    }

    vector<Edge> mst;
    int totalWeight = 0;

    while (!pq.empty() && visited.size() < vertices.size()) {
      auto [w, u, v] = pq.top();

      pq.pop();

      if (visited.find(v) != visited.end()) {
        continue;
      }

      visited.insert(v);
      mst.push_back({w, u, v});
      totalWeight += w;

      for (auto &p : nodes[v].adj) {
        if (visited.find(p.first) == visited.end()) {
          pq.push({p.second, v, p.first});
        }
      }
    }

    cout << "\nMST (Prim's):\n";

    for (auto &[w, u, v] : mst) {
      cout << u << " -- " << v << " weight = " << w << "\n";
    }

    cout << "Total Weight: " << totalWeight << "\n\n";
  }

  void Kruskal() {
    unordered_map<string, string> parent;
    unordered_map<string, int> rank;

    for (auto &v : vertices) {
      parent[v] = v;
      rank[v] = 0;
    }

    using Edge = tuple<int, string, string>;
    vector<Edge> sortedEdges = edges;

    sort(sortedEdges.begin(), sortedEdges.end());

    vector<Edge> mst;
    int totalWeight = 0;

    for (auto &[w, u, v] : sortedEdges) {
      string ru = findParent(u, parent);
      string rv = findParent(v, parent);

      if (ru != rv) {
        unionSet(ru, rv, parent, rank);
        mst.push_back({w, u, v});
        totalWeight += w;
      }
    }

    cout << "\nMST (Kruskal's):\n";

    for (auto &[w, u, v] : mst) {
      cout << u << " -- " << v << " weight = " << w << "\n";
    }

    cout << "Total Weight: " << totalWeight << "\n\n";
  }

  void ShortestPath(string &start) {
    if (nodes.find(start) == nodes.end()) {
      cout << "Invalid start vertex\n";
      return;
    }

    unordered_map<string, int> dist;
    unordered_map<string, string> parent;

    for (auto &v : vertices) {
      dist[v] = INT_MAX;
      parent[v] = "";
    }

    dist[start] = 0;

    using P = pair<int, string>;
    priority_queue<P, vector<P>, greater<P>> pq;

    pq.push({0, start});

    while (!pq.empty()) {
      auto [d, u] = pq.top();

      pq.pop();

      if (d > dist[u]) {
        continue;
      }

      for (auto &p : nodes[u].adj) {
        int newDist = d + p.second;

        if (newDist < dist[p.first]) {
          dist[p.first] = newDist;
          parent[p.first] = u;
          pq.push({newDist, p.first});
        }
      }
    }

    cout << "\nShortest paths from " << start << ":\n";

    for (auto &v : vertices) {
      if (dist[v] == INT_MAX) {
        cout << start << " -> " << v << ": No path\n";
      } else {
        vector<string> path;

        for (string node = v; !node.empty(); node = parent[node]) {
          path.insert(path.begin(), node);
        }

        cout << start << " -> " << v << ": Distance = " << dist[v]
             << ", Path = ";

        for (size_t i = 0; i < path.size(); i++) {
          cout << path[i];

          if (i != path.size() - 1) {
            cout << " -> ";
          }
        }

        cout << "\n";
      }
    }
  }

  void printGraph() {
    cout << "\nGraph Adjacency List:\n";

    for (auto &kv : nodes) {
      cout << kv.first << ": ";

      for (auto &p : kv.second.adj) {
        cout << "(" << p.first << ", " << p.second << ") ";
      }

      cout << "\n";
    }

    cout << "\n";
  }
};

int main() {
  Graph graph;

  int inputs;
  cin >> inputs;
  cin.ignore();

  while (inputs-- > 0) {
    string cmd;
    getline(cin, cmd);
    vector<string> op;
    string temp;

    for (auto c : cmd) {
      if (c == ' ') {
        if (!temp.empty()) {
          op.push_back(temp);
        }

        temp.clear();
      } else {
        temp += c;
      }
    }

    if (!temp.empty()) {
      op.push_back(temp);
    }

    if (op[0] == "IN") {
      graph.insertNode(op[1]);
    } else if (op[0] == "IE") {
      graph.insertEdge(op[1], op[2], stoi(op[3]));
      graph.printGraph();
    } else if (op[0] == "DFS") {
      graph.DFS(op[1]);
    } else if (op[0] == "BFS") {
      graph.BFS(op[1]);
    } else if (op[0] == "PRIM") {
      graph.Prim(op[1]);
    } else if (op[0] == "KRUSKAL") {
      graph.Kruskal();
    } else if (op[0] == "DIJKSTRA") {
      graph.ShortestPath(op[1]);
    }

    return 0;
  }
}

File: data-structures/non-linear/graph/Graph.py
# Uses MinHeap ADT from data-structures/non-linear/binary-tree/heap/min-heap/MinHeap.py


class Graph:
    class Node:
        def __init__(self, e):
            self.AList = []  # neighbors (neighbor_value, weight)
            self.value = e

        def addtoAList(self, neighbor, w):
            self.AList.append((neighbor, w))

    class Edge:
        def __init__(self, v1, v2, w):
            self.v1 = v1
            self.v2 = v2
            self.weight = w

    def __init__(self):
        self.nodes = {}  # key: node value, value: Node
        self.edges = []  # list of edges (weight, v1, v2)
        self.vertices = set()

    def insertNode(self, e):
        if e not in self.nodes:
            self.nodes[e] = self.Node(e)
            self.vertices.add(e)

    def insertEdge(self, v1, v2, w):
        if v1 not in self.nodes or v2 not in self.nodes:
            print("Vertex not found")
            return
        for n, weight in self.nodes[v1].AList:
            if n == v2:
                print(f"Edge {v1}-{v2} already exists")
                return
        self.nodes[v1].addtoAList(v2, w)
        self.nodes[v2].addtoAList(v1, w)

        edge = self.Edge(v1, v2, w)
        self.edges.append(edge)

    def getNode(self, e):
        return self.nodes.get(e, None)

    def DFS(self, v, visited=None):
        if v is None:
            return
        if visited is None:
            visited = set()
        visited.add(v.value)
        print(v.value, end=" ")
        for neighbor, _ in v.AList:
            if neighbor not in visited:
                self.DFS(self.nodes[neighbor], visited)

    def BFS(self, v):
        if v is None:
            return
        visited = set()
        q = [v]
        visited.add(v.value)
        while q:
            node = q.pop(0)
            print(node.value, end=" ")
            for neighbor, _ in node.AList:
                if neighbor not in visited:
                    visited.add(neighbor)
                    q.append(self.nodes[neighbor])

    def MST_Prim(self, start):
        if start not in self.nodes:
            print("Invalid start vertex")
            return
        visited = set([start])
        heap = MinHeap()
        mst, total_weight = [], 0
        for neighbor, w in self.nodes[start].AList:
            heap.insert((w, start, neighbor))
        while heap.size() > 0 and len(visited) < len(self.vertices):
            w, u, v = heap.extract_min()
            if v not in visited:
                visited.add(v)
                mst.append((u, v, w))
                total_weight += w
                for n, wt in self.nodes[v].AList:
                    if n not in visited:
                        heap.insert((wt, v, n))
        print("\nMST (Prim's):")
        for u, v, w in mst:
            print(f"{u} -- {v}  weight = {w}")
        print("Total Weight:", total_weight, "\n")

    def MST_Kruskal(self):
        parent = {}
        rank = {}

        def find(v):
            if parent[v] != v:
                parent[v] = find(parent[v])
            return parent[v]

        def union(u, v):
            root1 = find(u)
            root2 = find(v)
            if root1 != root2:
                if rank[root1] < rank[root2]:
                    parent[root1] = root2
                elif rank[root1] > rank[root2]:
                    parent[root2] = root1
                else:
                    parent[root2] = root1
                    rank[root1] += 1

        for node in self.nodes:
            parent[node] = node
            rank[node] = 0

        heap = MinHeap([(edge.weight, edge.v1, edge.v2) for edge in self.edges])
        sorted_edges = [heap.extract_min() for _ in range(len(self.edges))]

        MST = []
        total_weight = 0

        for w, u, v in sorted_edges:
            if find(u) != find(v):
                union(u, v)
                MST.append((u, v, w))
                total_weight += w

        print("\nKruskal's MST:")
        for u, v, w in MST:
            print(f"{u} - {v} : {w}")
        print("Total Weight:", total_weight)

    def ShortestPath(self, start):
        if start not in self.nodes:
            print("Invalid start vertex")
            return
        distances = {v: float("inf") for v in self.vertices}
        parent = {v: None for v in self.vertices}
        distances[start] = 0
        visited = set()
        heap = MinHeap([(0, start)])
        while heap.size() > 0:
            dist, u = heap.extract_min()
            if u in visited:
                continue
            visited.add(u)
            for neighbor, w in self.nodes[u].AList:
                new_dist = dist + w
                if new_dist < distances[neighbor]:
                    distances[neighbor] = new_dist
                    parent[neighbor] = u
                    heap.insert((new_dist, neighbor))
        print(f"\nShortest paths from {start}:")
        for v in self.vertices:
            if distances[v] == float("inf"):
                print(f"{start} → {v}: No path")
            else:
                path = []
                node = v
                while node is not None:
                    path.insert(0, node)
                    node = parent[node]
                print(
                    f"{start} → {v}: Distance = {distances[v]}, Path = {' -> '.join(path)}"
                )

    def printGraph(self):
        print("\nGraph Adjacency List:")
        for key, node in self.nodes.items():
            print(f"{key}: {node.AList}")
        print("")


def main():
    graph = Graph()
    inputs = int(input())
    while inputs > 0:
        command = input()
        operation = command.split()
        if operation[0] == "IN":
            graph.insertNode(operation[1])
        elif operation[0] == "IE":
            graph.insertEdge(operation[1], operation[2], int(operation[3]))
            graph.printGraph()
        elif operation[0] == "DFS":
            graph.DFS(graph.getNode(operation[1]))
            print("")
        elif operation[0] == "BFS":
            graph.BFS(graph.getNode(operation[1]))
            print("")
        elif operation[0] == "PRIM":
            graph.MST_Prim(operation[1])
        elif operation[0] == "KRUSKAL":
            graph.MST_Kruskal()
        elif operation[0] == "DIJKSTRA":
            graph.ShortestPath(operation[1])
        inputs -= 1


if __name__ == "__main__":
    main()
